# 1. Getting The Data

In [20]:
import numpy as np 
import pandas as pd 
data=pd.read_csv("data.csv",encoding='cp1252')
pd.set_option('display.max_columns',None) # display all the features 
data.head(5)

C:\Users\darka\AppData\Local\Temp\ipykernel_14268\3846487868.py:3: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  data=pd.read_csv("data.csv",encoding='cp1252')


stn_code       sampling_date           state   location agency  \
0    150.0  February - M021990  Andhra Pradesh  Hyderabad    NaN   
1    151.0  February - M021990  Andhra Pradesh  Hyderabad    NaN   
2    152.0  February - M021990  Andhra Pradesh  Hyderabad    NaN   
3    150.0     March - M031990  Andhra Pradesh  Hyderabad    NaN   
4    151.0     March - M031990  Andhra Pradesh  Hyderabad    NaN   

                                 type  so2   no2  rspm  spm  \
0  Residential, Rural and other Areas  4.8  17.4   NaN  NaN   
1                     Industrial Area  3.1   7.0   NaN  NaN   
2  Residential, Rural and other Areas  6.2  28.5   NaN  NaN   
3  Residential, Rural and other Areas  6.3  14.7   NaN  NaN   
4                     Industrial Area  4.7   7.5   NaN  NaN   

  location_monitoring_station  pm2_5        date  
0                         NaN    NaN  1990-02-01  
1                         NaN    NaN  1990-02-01  
2                         NaN    NaN  1990-02-01  
3                         NaN    NaN  1990-03-01  
4                         NaN    NaN  1990-03-01

# 2. Analysing the Data

Data Information

In [93]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 435742 entries, 0 to 435741
Data columns (total 13 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   stn_code                     291665 non-null  object 
 1   sampling_date                435739 non-null  object 
 2   state                        435742 non-null  object 
 3   location                     435739 non-null  object 
 4   agency                       286261 non-null  object 
 5   type                         430349 non-null  object 
 6   so2                          401096 non-null  float64
 7   no2                          419509 non-null  float64
 8   rspm                         395520 non-null  float64
 9   spm                          198355 non-null  float64
 10  location_monitoring_station  408251 non-null  object 
 11  pm2_5                        9314 non-null    float64
 12  date                         435735 non-null  object 
dtyp

- Dataset contains 13 features, where there 8 features in object type and 5 features in float64 type.

In [94]:
data.describe ()

so2            no2           rspm            spm        pm2_5
count  401096.000000  419509.000000  395520.000000  198355.000000  9314.000000
mean       10.829414      25.809623     108.832784     220.783480    40.791467
std        11.177187      18.503086      74.872430     151.395457    30.832525
min         0.000000       0.000000       0.000000       0.000000     3.000000
25%         5.000000      14.000000      56.000000     111.000000    24.000000
50%         8.000000      22.000000      90.000000     187.000000    32.000000
75%        13.700000      32.200000     142.000000     296.000000    46.000000
max       909.000000     876.000000    6307.033333    3380.000000   504.000000

## Checking Duplicates Values

In [95]:
print(data.duplicated().value_counts())
data.drop_duplicates(inplace = True)
print(len(data))

False    435068
True        674
dtype: int64
435068


## Checking Missing Values

In [96]:
print('Data columns with null values:\n',
     data.isnull().sum())

Data columns with null values:
 stn_code                       144077
sampling_date                       3
state                               0
location                            3
agency                         149466
type                             5357
so2                             34632
no2                             16222
rspm                            40035
spm                            236908
location_monitoring_station     27303
pm2_5                          425754
date                                7
dtype: int64


* We can see that there are many null values present in feature:stn_code, agency, type, so2, no2, rspm, spm, location_monitoring_station, pm2_5

## Checking Unique Values


In [97]:
data.nunique()

stn_code                        803
sampling_date                  5485
state                            37
location                        304
agency                           64
type                             10
so2                            4197
no2                            6864
rspm                           6065
spm                            6668
location_monitoring_station     991
pm2_5                           433
date                           5067
dtype: int64

- From the above output we can have an overall idea of how many unique values present in the dataset. 

# 3. Data Filteration & Modification

## Dropping unnessesary column(s)

* stn_code, agency & sampling_date  this two features is irrelevant to our analysis so let's remove this feature

In [21]:
data=data.drop(['stn_code'],axis=1)
data=data.drop(['agency'],axis=1)
data=data.drop(['sampling_date'],axis=1)

* We are replacing it all the types of locations and getting them down to only 4 following categories:

   * Residential Area

   * Industrial Area

   * Sensitive Area

   * RIRUO 

In [22]:
data['type'].replace('Residential, Rural and other Areas','Residential Area',inplace = True)
data['type'].replace('Residential','Residential Area',inplace = True)
data['type'].replace('Residential and others','Residential Area',inplace = True)
data['type'].replace('Industrial Areas','Industrial Area',inplace = True)
data['type'].replace('Industrial','Industrial Area',inplace = True)
data['type'].replace('Sensitive','Sensitive Area',inplace = True)
data['type'].replace('Sensitive Areas','Sensitive Area',inplace = True)

data['type'].value_counts()

Residential Area    265963
Industrial Area     148071
Sensitive Area       15011
RIRUO                 1304
Name: type, dtype: int64

In [23]:
data.head()

state   location              type  so2   no2  rspm  spm  \
0  Andhra Pradesh  Hyderabad  Residential Area  4.8  17.4   NaN  NaN   
1  Andhra Pradesh  Hyderabad   Industrial Area  3.1   7.0   NaN  NaN   
2  Andhra Pradesh  Hyderabad  Residential Area  6.2  28.5   NaN  NaN   
3  Andhra Pradesh  Hyderabad  Residential Area  6.3  14.7   NaN  NaN   
4  Andhra Pradesh  Hyderabad   Industrial Area  4.7   7.5   NaN  NaN   

  location_monitoring_station  pm2_5        date  
0                         NaN    NaN  1990-02-01  
1                         NaN    NaN  1990-02-01  
2                         NaN    NaN  1990-02-01  
3                         NaN    NaN  1990-03-01  
4                         NaN    NaN  1990-03-01

# Handling Null Values

* As we have seen already that in some features the null values are very few so we can remove those rows directly, it will hadrly effect our overall analysis. Below are the features that we are removing:
 
 1. location
 2. date

* And also we have seen that there few feature columns having a huge chunk on missing values. By dropping those rows from tha dataset will negetively effect our further analysis, so instead of removing those null values we are going to replace them with the "mean value". Below are the features that we are going to replace values:

 1. so2
 2. no2
 3. rspm
 4. spm
 5. pm2_5

In [24]:
data.dropna(subset=['location'],how='any',inplace=True)     # droping null value rows of "location" column
data.dropna(subset=['date'],how='any',inplace=True)   # droping null value rows of "date" column

In [25]:
# mean values of so2,no2,rspm,spm,pm2_5
print("The mean values of so2,no2,rspm,spm,pm2_5\n")
meanVal_1 = data['so2'].mean()
meanVal_2 = data['no2'].mean()
meanVal_3 = data['rspm'].mean()
meanVal_4 = data['spm'].mean()
meanVal_5 = data['pm2_5'].mean()
print("meanVal_1 - so2 : ", meanVal_1)
print("meanVal_2 - no2 : ", meanVal_2)
print("meanVal_3 - rspm : ", meanVal_3)
print("meanVal_4 - spm : ", meanVal_4)
print("meanVal_4 - pm2_5 : ", meanVal_5)

The mean values of so2,no2,rspm,spm,pm2_5

meanVal_1 - so2 :  10.829427580620525
meanVal_2 - no2 :  25.809659222747314
meanVal_3 - rspm :  108.83309095208052
meanVal_4 - spm :  220.78347959832885
meanVal_4 - pm2_5 :  40.791467038866244


In [26]:
data['so2'].fillna(value=meanVal_1, inplace=True)  # replacing NaN value with "mean value"
data['no2'].fillna(value=meanVal_2, inplace=True)       # replacing NaN value with "mean value"
data['rspm'].fillna(value=meanVal_3, inplace=True)    # replacing NaN value with "mean value"
data['spm'].fillna(value=meanVal_4, inplace=True)    # replacing NaN value with "mean value"
data['pm2_5'].fillna(value=meanVal_5, inplace=True)    # replacing NaN value with "mean value"

In [27]:
data.head()

state   location              type  so2   no2        rspm  \
0  Andhra Pradesh  Hyderabad  Residential Area  4.8  17.4  108.833091   
1  Andhra Pradesh  Hyderabad   Industrial Area  3.1   7.0  108.833091   
2  Andhra Pradesh  Hyderabad  Residential Area  6.2  28.5  108.833091   
3  Andhra Pradesh  Hyderabad  Residential Area  6.3  14.7  108.833091   
4  Andhra Pradesh  Hyderabad   Industrial Area  4.7   7.5  108.833091   

         spm location_monitoring_station      pm2_5        date  
0  220.78348                         NaN  40.791467  1990-02-01  
1  220.78348                         NaN  40.791467  1990-02-01  
2  220.78348                         NaN  40.791467  1990-02-01  
3  220.78348                         NaN  40.791467  1990-03-01  
4  220.78348                         NaN  40.791467  1990-03-01

## Now let's check again if any null values are present or not.

In [28]:
print('Data columns with null values:\n',
     data.isnull().sum())

Data columns with null values:
 state                              0
location                           0
type                            5390
so2                                0
no2                                0
rspm                               0
spm                                0
location_monitoring_station    27488
pm2_5                              0
date                               0
dtype: int64


* So now there is no null values in so2, no2, rspm, spm, pm2_5 and now we can calculate AQI for the above dataset

# 4. Air Quality Index

* AQI: The air quality index is an index for reporting air quality on a daily basis.  In other words, it is a measure of how air pollution affects one’s health within a short time period. The AQI is calculated based on the average concentration of a particular pollutant measured over a standard time interval. Generally, the time interval is 24 hours for most pollutants, 8 hours for carbon monoxide and ozone.

* We can see how air pollution is by looking at the AQI level:
  * Good(0 – 50)
  * Moderate(51 – 100)
  * Unhealthy(101 – 150)
  * Unhealthy for Strong People(151 – 200)
  * Hazardous(201+)

In [30]:
#Function to calculate so2 individual pollutant index(si)
def calculate_si(so2):
    si=0
    if (so2<=40):
        si= so2*(50/40)
    elif (so2>40 and so2<=80):
        si= 50+(so2-40)*(50/40)
    elif (so2>80 and so2<=380):
        si= 100+(so2-80)*(100/300)
    elif (so2>380 and so2<=800):
        si= 200+(so2-380)*(100/800)
    elif (so2>800 and so2<=1600):
        si= 300+(so2-800)*(100/800)
    else:
        si= 400+(so2-1600)*(100/800)
    return si
data['si']=data['so2'].apply(calculate_si)
df= data[['so2','si']]
df.head()

so2     si
0  4.8  6.000
1  3.1  3.875
2  6.2  7.750
3  6.3  7.875
4  4.7  5.875

In [31]:
#Function to calculate no2 individual pollutant index(ni)
def calculate_ni(no2):
    ni=0
    if(no2<=40):
        ni= no2*50/40
    elif(no2>40 and no2<=80):
        ni= 50+(no2-14)*(50/40)
    elif(no2>80 and no2<=180):
        ni= 100+(no2-80)*(100/100)
    elif(no2>180 and no2<=280):
        ni= 200+(no2-180)*(100/100)
    elif(no2>280 and no2<=400):
        ni= 300+(no2-280)*(100/120)
    else:
        ni= 400+(no2-400)*(100/120)
    return ni
data['ni']=data['no2'].apply(calculate_ni)
df= data[['no2','ni']]
df.head()

no2      ni
0  17.4  21.750
1   7.0   8.750
2  28.5  35.625
3  14.7  18.375
4   7.5   9.375

In [32]:
#Function to calculate rspm individual pollutant index(rpi)
def calculate_(rspm):
    rpi=0
    if(rpi<=30):
        rpi=rpi*50/30
    elif(rpi>30 and rpi<=60):
        rpi=50+(rpi-30)*50/30
    elif(rpi>60 and rpi<=90):
        rpi=100+(rpi-60)*100/30
    elif(rpi>90 and rpi<=120):
        rpi=200+(rpi-90)*100/30
    elif(rpi>120 and rpi<=250):
        rpi=300+(rpi-120)*(100/130)
    else:
        rpi=400+(rpi-250)*(100/130)
    return rpi
data['rpi']=data['rspm'].apply(calculate_si)
df= data[['rspm','rpi']]
df.tail()

rspm         rpi
435734  148.0  122.666667
435735  131.0  117.000000
435736  140.0  120.000000
435737  143.0  121.000000
435738  171.0  130.333333

In [33]:
#Function to calculate spm individual pollutant index(spi)
def calculate_spi(spm):
    spi=0
    if(spm<=50):
        spi=spm
    elif(spm<50 and spm<=100):
        spi=spm
    elif(spm>100 and spm<=250):
        spi= 100+(spm-100)*(100/150)
    elif(spm>250 and spm<=350):
        spi=200+(spm-250)
    elif(spm>350 and spm<=450):
        spi=300+(spm-350)*(100/80)
    else:
        spi=400+(spm-430)*(100/80)
    return spi
data['spi']=data['spm'].apply(calculate_spi)
df= data[['spm','spi']]
df.tail()

spm        spi
435734  220.78348  180.52232
435735  220.78348  180.52232
435736  220.78348  180.52232
435737  220.78348  180.52232
435738  220.78348  180.52232

In [34]:
#function to calculate the air quality index (AQI) of every data value
def calculate_aqi(si,ni,spi,rpi):
    aqi=0
    if(si>ni and si>spi and si>rpi):
        aqi=si
    if(spi>si and spi>ni and spi>rpi):
        aqi=spi
    if(ni>si and ni>spi and ni>rpi):
        qi=ni
    if(rpi>si and rpi>ni and rpi>spi):
        aqi=rpi
    return aqi
data['AQI']=data.apply(lambda x:calculate_aqi(x['si'],x['ni'],x['spi'],x['rpi']),axis=1)
df= data[['date','state','si','ni','rpi','spi','AQI']]
df.head()

date           state     si      ni        rpi        spi        AQI
0  1990-02-01  Andhra Pradesh  6.000  21.750  109.61103  180.52232  180.52232
1  1990-02-01  Andhra Pradesh  3.875   8.750  109.61103  180.52232  180.52232
2  1990-02-01  Andhra Pradesh  7.750  35.625  109.61103  180.52232  180.52232
3  1990-03-01  Andhra Pradesh  7.875  18.375  109.61103  180.52232  180.52232
4  1990-03-01  Andhra Pradesh  5.875   9.375  109.61103  180.52232  180.52232

# 5. Exporting the Purified Data

In [18]:
data.to_csv('data_cleaned.csv')